make __doc__ of class members writable

N-Coder opened the pull request:
See my initial issue [here](https://bitbucket.org/wlav/cppyy/issues/339/make-c-documentation-available-to-python). Regarding your [comment](https://bitbucket.org/wlav/cppyy/issues/369/template-instantiation-not-happening-in#comment-61161603):
> I was kindof hoping to find a solution that doesn’t require adding another pointer CPPOverload::MethodInfo_t. Not sure whether that’s possible, though, but the whole structure could be reduced: there need not be a separate dispatch map as a loop over methods would be just as fast and fName is hardly used (only in error reporting and __str__), so could be on-demand instead.

...I took the easy path and did exactly that to `CPPDataMember` (where I could add the member field directly to the class), `TemplateProxy`, and `CPPOverload` (where I added the member field to the persistent `CPPOverload::MethodInfo_t` / `TemplateInfo` object and adapted the `__doc__` getter of / added a setter to the enclosing object). By default, the getters return the auto-generated docstring, unless the user overwrote it with a custom value. I guess you need to store the string value somewhere, unless you want to give the objects their whole, customizable `__dict__` with `tp_dictoffset`, but you'd also need to store that somewhere. I'm unsure whether something similar needs to be done to `CustomInstanceMethod`, `TypedefPointerToClass`, `CPPInstance`, and `CPPExcInstance`, as I didn't have a class at hand that wraps those. At least the current changes in this PR seem to cover my use-case from the Bitbucket issue.

Regarding the hook for `CPPScope.__getattr__` discussed [here](https://bitbucket.org/wlav/cppyy/issues/339/make-c-documentation-available-to-python#comment-60495916), I guess I would also add a PyCallable member to `CPPScope` that gets called [here](https://github.com/wlav/CPyCppyy/blob/master/src/CPPScope.cxx#L518) in the native implementation of `meta_getattro` to provide a replacement value if no other alternative could be found.

wlav commented on src/CPPOverload.h:
```diff
@@ -40,13 +40,15 @@ class CPPOverload {
         CPPOverload::DispatchMap_t  fDispatchMap;
         CPPOverload::Methods_t      fMethods;
         uint32_t                    fFlags;
+        PyObject*                   fDoc;
```
Recommend to place `fDoc` above `fFlags`. The latter being half a word size on 64b platforms, so placing the smallest data members last may, in some cases, save on alignment.

wlav commented:
Yes, storage somewhere is needed. The only alternative that I can think of, is to use an "extended" info, same as done for `CPPInstance`. Problem is how to update the outstanding pointers, otherwise both versions (with and without) would be around for a while. One solution would be to capture other little-used or on-demand constructable pieces (e.g. `fName` and `fDescription`) and put them together with `fDoc` in an on-demand extension. Note that if only one is ever used, the pointer to the extension can be the pointer to the current use. Only if both need constructing is an extension needed.

`CPPInstance` actually has a `__dict__` and `CPPExcInstance` would be ephemeral enough not to need `__doc__`. In both cases, the better place would be in `CPPScope`. `CustomInstanceMethod` would likewise have the problem of being ephemeral (`CPPOverload` and `TemplateProxy` differ here: their innards are not only shared for reduced memory usage, but also for settings to survive if assigned on a temporary, e.g. bound, copy). `TypedefPointerToClass` is fine to grow  a `__dict__`: it's an uncommon beast.

N-Coder commented on src/CPPDataMember.cxx:
```diff
@@ -176,6 +177,7 @@ static CPPDataMember* pp_new(PyTypeObject* pytype, PyObject*, PyObject*)
     pyprop->fConverter      = nullptr;
     pyprop->fEnclosingScope = 0;
     pyprop->fDescription    = nullptr;
+    pyprop->fDoc            = PyUnicode_FromString("");
```
`CPPDataMember` is the only one to have [direct access](https://github.com/wlav/CPyCppyy/pull/1/files#diff-32dc08abe156fbe8ad6252c1210ad1c4a044c640420f94e781dece2d744d354aR197) to the doc field/member instead of having some customized behaviour via a getter. I'm not sure how Python handles C `nullptr` vs Python `None`, so I wanted to stay on the safe side. But maybe we could replace `T_OBJECT_EX` by `T_OBJECT` and then leave it at `nullptr` initially, depending on how well this plays with Python's `help`.

N-Coder replied:
fix ind 3d5e7471f6139a7d40154500ec62df73d21cfd72

wlav replied:
The empty string is not Py_None, so that's not the comparison. Point is rather that if any other code wants to distinguish between "set" and "not set", then some Python string v.s `nullptr` is an easy way to distinguish.

N-Coder replied:
I was mostly unsure how Python reacts to the `__doc__` field being `nullptr`, i.e. whether that is interpreted as `None`, the empty string, or raises an exception. But it seems to work with setting it to `nullptr` initially. So is it alright the way it is now?

N-Coder commented:
My use-case usually doesn't involve memory pressure, so I'd be happy with the simple solution, but I guess a more involved implementation can always be added later on. :sweat_smile:

Do you have an example where `TypedefPointerToClass` occurs? Then I could add and test that.

N-Coder commented:
I just create a [small test case](https://gist.github.com/N-Coder/a7d63e0e2cc89c7062c75620e1d36802#file-test2-py) and found some oddities (scroll up in the gist to see the output):
- `CPPDataMember` 'field' is listed and accessible in `__dict__`, but not accessible via `inst.field` or `getattr(inst, "field")`
- `TemplateProxy` methods start out with the docstring "cppyy template proxy (internal)" instead of the signatures of the represented methods
- `TemplateProxy`s seem to use their `__doc__` string as output for `__repr__` instead of something like `<cppyy.TemplateProxy object at 0x7f1d57a15f00>` (see the dumped `__dict__`s at the very begining of log, you need to scroll horizontally to find `templ`)
- `help()` prints the `__doc__` of `TemplateProxy`s twice, but that might be due to the previous point:
```
 |  templ = Description of the templ.
 |      Description of the templ.
 ```

Should I try to fix those points, too?

wlav commented:
> Do you have an example where `TypedefPointerToClass` occurs? Then I could add and test that.

Here's an example (it's a common one for opaque handles):

```
import cppyy
  
cppyy.cppdef("""\
namespace doctest {

struct A{};

struct B {
    typedef A* A_t;
}; }""")

ns = cppyy.gbl.doctest

a_ptr = ns.B.A_t(0x2a)

assert cppyy.addressof(a_ptr) == 42
```

And yes, `TemplateProxy` is a mess, but that's b/c it can grow (as more methods are instantiated) and it also represents non-templated overloads, specializations, etc. None of these need to have the same signature. And yes `__repr__` calls `tpp_doc`. Same reason: not much better to print in `repr()` as there isn't really a reconstructable state.

N-Coder commented:
Okay, I implemented that and updated the testing example. There are two things I'd still like to see fixed, afterwards this PR should be good to go. Have a look at the example `__dict__`:
```python
<class cppyy.gbl.Test at 0x55b79cb89e70> {'__module__': '__main__',
    '__dict__': <attribute '__dict__' of 'Test' objects>,
    '__weakref__': <attribute '__weakref__' of 'Test' objects>,
    '__doc__': None,
    'templ': cppyy template proxy (internal),
    '__assign__': <cppyy.CPPOverload object at 0x7f7673be1580>,
    '__init__': <cppyy.CPPOverload object at 0x7f7673bda840>,
    'method': <cppyy.CPPOverload object at 0x7f7673bda8c0>,
    'overload': <cppyy.CPPOverload object at 0x7f7673bdd800>,
    'field': <cppyy.CPPDataMember object at 0x7f7673bdd830>}
```
Here, all reprs say `<cppyy.Something object at 0x123456789abcdef>`, except for `TemplateProxy`, which prints its `__doc__`.
I'd unify this so that `TemplateProxy` uses the same repr as all the other objects.

As a second point, `field` is listed in the `__dict__` output, but calling `cppyy.gbl.Test.field` will throw a not-found exception. To access the attribute, you need to use `cppyy.gbl.Test.__dict__["field"]`, which is weird. I'm not quite sure how to fix this, but it's probably somewhere in `meta_getattro`.

N-Coder commented:
Python unfortunately has no built-in support for function overloading. There is only a `@singledispatch` decorator that differentiates based on the type of the first argument, but this decorator doesn't provide any `__annotations__` or other runtime typing information about the overloaded implementations. The [`@overload`](https://www.python.org/dev/peps/pep-0484/#function-method-overloading) [decorator](https://docs.python.org/3/library/typing.html#typing.overload) can unfortunately only be used in type stub files to tell type-checkers using non-`__annotations__` based analysis (i.e. directly parsing the stub file, not via live runtime object inspection) that a function accepts different pairs of argument types. I also couldn't find any other example where this kind of information was exposed during runtime via `__annotations__`.
So I see two (mutually non-exclusive) ways of going forward:
- Fill the `__annotations__` in a best-effort kind of way, e.g. throwing overloaded arguments into a `Union` and marking arguments not present in all overloads a `Optional`. This will be far from perfect, but should already help a lot in the most common cases and where runtime type inspection is used, e.g. in ipython shells and jupyter notebooks.
- Provide some other interface for obtaining the more detailled overload information so that e.g. the respective stub files could be generated automatically.

But maybe this is something for another issue and we should try to get the `__doc__` changes in first.

wlav commented:
Yes, I'm thinking the same: the only problem that my pop up is confusion, so perhaps limiting "best effort" to cases that are clear maybe enough.

I'm going to merge this PR later today and fix up the remaining issue (data members) directly in the code. I find that easier than discussing on a PR. :) I'm really hoping to get my Windows VMWare up and running again so that i can cut a release (last one was from July ...).
